### PR TITLE
Generic Part1

### PR DESCRIPTION
**Pull request by burrbull**
Rebased and reopened.

### CODE REVIEW

**Comment by burrbull** [comment]
cc @therealprof 

With this PR debug builds grow ~1.5% (I think it is because `Deref`), but source now much cleaner.

It also depends on #358.

**Comment by burrbull** [comment]
cc @Disasm

**Comment by burrbull** [comment]
@therealprof  I've rebased this on master. Test, please.

File: src/generate/generic.rs [hunk 1]
```diff
@@ -18,23 +30,97 @@ pub fn render() -> Result<Vec<Tokens>> {
...
+                self.register.set(f(&mut W {bits: Self::reset_value(), _reg: marker::PhantomData}).bits);
+            }
+        }
+    });
+
+    generic_items.push(quote! {
+        impl<U, REG> Reg<U, REG>
+        where
+            Self: Writable,
+            U: Copy + Default
```
Comment (Disasm) [inline src/generate/generic.rs]: Is this `Default` equals to "returns zero"?

**Comment by Disasm** [comment]
`Send` and `!Sync` were implemented on registers before, with this change they are not implemented.

File: src/generate/generic.rs (hunk 1 above)
Comment (burrbull) [reply src/generate/generic.rs]: Yes, it is for `bool`, `u8`, `u16` `u32` and `u64`.

**Comment by burrbull** [comment]
> `Send` and `!Sync` were implemented on registers before, with this change they are not implemented.

It was auto-implementation?

File: src/generate/generic.rs (hunk 1 above)
Comment (Disasm) [reply src/generate/generic.rs]: Yes, but these types are not enforced I guess

**Comment by Disasm** [comment]
Yes, I think that was due to the inner `VolatileCell`: it's also `Send` and `!Sync`.

**Comment by Disasm** [comment]
As discussed on Matrix, documentation becomes even more difficult to understand with this change: it's unclear how to read and write registers, also there is no information about field variants.

**Comment by therealprof** [comment]
TBH I'm not too sure whether we really need that `VolatileCell` anyway once we've remodelled everything. 😅

**Comment by Disasm** [comment]
@therealprof Yes, but it's better to have `Send` and `!Sync` anyway, because they add important properties.

**Comment by burrbull** [comment]
error[E0658]: negative trait bounds are not yet fully implemented; use marker types for now

What to do with this?

**Comment by Disasm** [comment]
I think that PhantomData with some weird type can help

**Comment by burrbull** [comment]
Can you write how you see implementation for `!Sync`?

**Comment by Disasm** [comment]
Seems like `!Sync` is already implemented for `Reg`, but as for `Send` it's difficult to say whether it's implemented or not for a specific register.

**Comment by burrbull** [comment]
I've added `Send` and slightly improved docs.

File: src/generate/generic.rs (hunk 1 above)
Comment (burrbull) [reply src/generate/generic.rs]: They are enforced by size type of register or field.

**Comment by burrbull** [comment]
What you mean `there is no information about field variants`?

**Comment by Disasm** [comment]
@burrbull It's no longer accessible from the register documentation pages. Only on upper levels.

File: src/generate/generic.rs (hunk 1 above)
Comment (Disasm) [reply src/generate/generic.rs]: But it's possible to create a wrapper with the same size and a custom Default impl that returns reset value, for example

Comment (burrbull) [reply src/generate/generic.rs]: How? All fields are private. You can get register only from peripheral as it was before.

**Comment by burrbull** [comment]
> @burrbull It's no longer accessible from the register documentation pages. Only on upper levels.

Can you show example?

**Comment by Disasm** [comment]
[spi0::RegisterBlock](https://docs.rs/k210-pac/0.1.0/k210_pac/spi0/struct.RegisterBlock.html) -> [CTRLR0](https://docs.rs/k210-pac/0.1.0/k210_pac/spi0/struct.CTRLR0.html) -> [R](https://docs.rs/k210-pac/0.1.0/k210_pac/spi0/ctrlr0/struct.R.html) -> [WORK_MODER](https://docs.rs/k210-pac/0.1.0/k210_pac/spi0/ctrlr0/enum.WORK_MODER.html)

**Comment by burrbull** [comment]
Here is one breaking change (#354). Reader now have `variant` method:
изображение

same as writer:

изображение

And now you can use this enum for both read and write:
```rust
let v = spi.ctrl0.read().work_mode().variant();
...
spi.ctrl0.write(|w| w.work_mode().variant(v)); // Before now incompatible types error was here.
```

**Comment by burrbull** [comment]
@therealprof I can't find where is source of this documentation https://docs.rs/svd2rust ?

File: src/generate/generic.rs (hunk 1 above)
Comment (Disasm) [reply src/generate/generic.rs]: Hmm, now I see. Thanks.

**Comment by therealprof** [comment]
@burrbull You mean `src/lib.rs`?

File: src/generate/generic.rs [hunk 2]
```diff
@@ -10,6 +10,27 @@ pub fn render() -> Result<Vec<Tokens>> {
...
+
+        ///Reset value of the register
+        ///
+        ///This value is initial value for `write` method.
+        ///It can be also directly writed to register by `reset` method.
+        pub trait ResetValue {
+            ///Register size
+            type Size;
+            ///Reset value of the register
+            fn reset_value() -> Self::Size;
```
Comment (Disasm) [inline src/generate/generic.rs]: This associated type name is debatable, but I'm not going to force my point of view here.

Comment (burrbull) [reply src/generate/generic.rs]: I've changed Size on Type

File: src/generate/register.rs [hunk 3]
```diff
@@ -210,12 +142,38 @@ pub fn render(
...
+        pub type #name_pc = crate::Reg<#rty, #_name_pc>;
+
+        #[allow(missing_docs)]
+        #[doc(hidden)]
+        pub struct #_name_pc;
+    });
 
+    let doc = format!("`read()` method returns [{0}::R]({0}::R) reader structure", &name_sc);
+    if can_read {
+        out.push(quote! {
```
Comment (therealprof) [inline src/generate/register.rs]: Why not move the `let doc = ...` inside the `if` clause? Should be cheaper to compute and avoids the risk of accidentally using it in a wrong context.

File: src/generate/register.rs [hunk 4]
```diff
@@ -210,12 +142,38 @@ pub fn render(
...
+            #[doc = #doc]
+            impl crate::Readable for #name_pc {}
+        });
+    }
+    let doc = format!("`write(|w| ..)` method takes [{0}::W]({0}::W) writer structure", &name_sc);
+    if can_write {
```
Comment (therealprof) [inline src/generate/register.rs]: Same here.

**Comment by therealprof** [comment]
Also please rebase and make sure the commits have a meaningful commit message.

**Comment by burrbull** [comment]
Rebased.

**Comment by therealprof** [comment]
bors try

**Comment by therealprof** [comment]
@Disasm Still happy?

**Comment by Disasm** [comment]
@therealprof yes, looks good.

**Comment by Disasm** [comment]
In general, I'd improve documentation for generic methods (read/write/...), replacing references to the book with short examples.

**Pull request merged**